### PR TITLE
WeightInfo for Multisig Pallet

### PR DESCRIPTION
**Pull request by shawntabrizi**
This introduces the automated benchmarking freamwork to the Multisig Pallet.

All extrinsic weights are now represented by a configurable `WeightInfo`.

### Before

```
34_000_000 + (2_000 * call_len) + call_weight
```

### After

```
14_749_000 + (1_000 * len) + call_weight
```

### Before

```
55_000_000 + (250_000 * sig_len) + (3_000 * call_len) + 2 Read + 1 Write
```

### After

```
74_556_000 + (135_000  * s ) + (1_000  * z) + 2 Read + 1 Write
```

### Before

```
55_000_000 + (250_000 * sig_len) + (3_000 * call_len) + 2 Read + 1 Write + 1 Write (store)
```

Missing `contains_key` read, so the new weight is correct

### After

```
90_218_000 + (129_000  * s ) + (3_000  * z) + 3 Read + 2 Write
```

### Before

```
55_000_000 + (250_000 * sig_len) + (3_000 * call_len) + 2 Read + 1 Write
```

Overweight by 1 read?

confirmed by https://www.shawntabrizi.com/substrate-graph-benchmarks/db-stats.html?p=multisig&e=as_multi_approve

### After

```
48_402_000 + (132_000 * sig_len) + (1_000 * call_len) + 1 Read + 1 Write
```

## cancel_as_multi

Old weight is overweight by 1 read and write due to not whitelisting

### CODE REVIEW

File: frame/multisig/src/benchmarking.rs [hunk 1]
```diff
@@ -161,8 +207,8 @@ benchmarks! {
 	approve_as_multi_approve {
 		// Signatories, need at least 2 people
 		let s in 2 .. T::MaxSignatories::get() as u32;
-		// Transaction Length
-		let z in 0 .. 10_000;
+		// Transaction Length not a component
```
Comment (apopiak) [inline frame/multisig/src/benchmarking.rs:210]:
```suggestion
		// Transaction Length, not a component
```

**Review by apopiak — APPROVED** [review]
LGTM, though the weight reductions are somewhat surprising to me (might be that I missed some changes to benchmarks?).

**Comment by shawntabrizi** [comment]
@apopiak i will make a larger companion with multiple weight files. Since this doesnt break the API, we should be good to go.

File: frame/multisig/src/lib.rs [hunk 2]
```diff
@@ -435,20 +374,18 @@ decl_module! {
...
-			true, // assume worst case: refunded
-		)]
+		#[weight = {
+			let s = other_signatories.len() as u32;
+
+			T::WeightInfo::approve_as_multi_create(s)
+			.max(T::WeightInfo::approve_as_multi_approve(s))
+			.max(T::WeightInfo::approve_as_multi_complete(s))
+			.saturating_add(*max_weight)
+		}]
```
Comment (thiolliere) [inline frame/multisig/src/lib.rs:388]: here the call is still decoded in some situation (like if it gets complete), thus decoding the code is still `O(call_len)` but anyway it should be small (like we don't hash it here). And it hasn't been introduced by this PR.

**Review by thiolliere — APPROVED** [review]
PR is good to me

**Comment by kianenigma** [comment]
Ain't the weight of `as_multi_threshold_1 ` dropped by half for no reason?

**Comment by shawntabrizi** [comment]
@kianenigma but results are consistent on the benchmark machine. I have nothing else to really go off of.

**Review by kianenigma — APPROVED** [review]

**Comment by gavofyork** [comment]
bot merge

**Pull request merged**